[ui] Token management interface on policy pages

philrenaud opened the pull request:
Adds the ability to view, delete, and create tokens associated with a given policy in the new [Policies UI](https://github.com/hashicorp/nomad/pull/13976).

Ability to do any of this is available only to `manager` tokens only.

![image](https://user-images.githubusercontent.com/713991/206574930-27ff40ea-6393-4414-9400-aa0f9fda53fc.png)

philrenaud commented on ui/mirage/factories/token.js:
```diff
@@ -14,6 +14,7 @@ export default Factory.extend({
   oneTimeSecret: () => faker.random.uuid(),
 
   afterCreate(token, server) {
+    if (token.policyIds && token.policyIds.length) return;
```
Don't generate mirage policyIDs if we're explicitly creating them elsewhere

philrenaud commented on ui/app/styles/components/policies.scss:
```diff
@@ -25,3 +25,48 @@ table.policies {
+  }
+}
+
+@media #{$mq-hidden-gutter} {
+  .token-operations {
+    grid-auto-flow: row;
+    grid-template-columns: 1fr;
+  }
+}
```
De-columnizes things if you're on a narrow screen
![image](https://user-images.githubusercontent.com/713991/206575283-d85e34aa-6de1-4e32-ad8e-f68e5ff3b8b4.png)

mikenomitch commented:
Feeling really good.

Small ignorable request: When you create an example token, do you think we could auto-copy to the clipboard? Seems marginally nicer than having the user copy from the alert.

Also, I think the policy name has to be wrapped in quotes in case they have a space. This command will fail for instance:
`nomad acl token create -name="<TOKEN_NAME>" -policy=back up -type=client -ttl=<8h>`

Should Policy Deleted and Token Deleted alerts be auto-removed? I see why the Example Token created one wouldn't be (since you have to copy), but it might be nice to remove the others after N seconds.

philrenaud commented:
@mikenomitch Added a "Copy to Clipboard" button and set timeout on that particular notification to 30sec.

Changed timeouts for token and policy deletion from 60sec (`¯\_(ツ)_/¯`) to 5sec

Added quotes to the example CLI command, thanks for spotting that!

ChaiWithJai commented on ui/app/abilities/token.js:
```diff
@@ -0,0 +1,10 @@
+import { alias } from '@ember/object/computed';
+
+export default class extends AbstractAbility {
+  @alias('selfTokenIsManagement') canRead;
+  @alias('selfTokenIsManagement') canList;
+  @alias('selfTokenIsManagement') canWrite;
+  @alias('selfTokenIsManagement') canUpdate;
+  @alias('selfTokenIsManagement') canDestroy;
```
`question`: Shouldn't these permissions be based on multiple properties?

philrenaud replied:
Token access and operations will only ever be available to Management tokens. If we want to have some kind of `policiesSupportTokenList` functionality later, it can be added to this file, but for now it's managers only — same as policies.

ChaiWithJai commented on ui/app/components/policy-editor.js:
```diff
@@ -22,11 +22,18 @@ export default class PolicyEditorComponent extends Component {
-        throw new Error(
-          'Policy name must be 1-128 characters long and can only contain letters, numbers, and dashes.'
-        );
+        throw {
+          errors: [
+            {
+              detail:
+                'Policy name must be 1-128 characters long and can only contain letters, numbers, and dashes.',
+            },
+          ],
+        };
       }
 
+      const shouldRedirectAfterSave = this.policy.isNew;
```
`question`: Looks we're shaping this error message to match the schema of an `AdapterError`. Are we just doing this because we don't want to update the utility function `messageError` that we have?

If `ember-data` updates the schema of the `AdapterError` then we'll have a pretty challenging situation to understand if we broke something.

Could we patch `messageError` to work with errors that don't stem from an HTTP response because I imagine we'll throw more errors in components because we've just recently started using flash messages?

philrenaud commented on ui/app/controllers/policies/policy.js:
```diff
@@ -43,4 +55,69 @@ export default class PoliciesPolicyController extends Controller {
+        type: 'client',
+      });
+      yield newToken.save();
+      this.refreshTokens();
```
Good catch; adding a yield!

philrenaud commented on ui/app/routes/policies/policy.js:
```diff
@@ -2,15 +2,23 @@ import Route from '@ember/routing/route';
-    return this.store.findRecord('policy', decodeURIComponent(params.name), {
-      reload: true,
+  async model(params) {
+    return await hash({
+      policy: this.store.findRecord('policy', decodeURIComponent(params.name), {
+        reload: true,
+      }),
+      tokens: this.store
+        .peekAll('token')
+        .filter((token) =>
+          token.policyNames?.includes(decodeURIComponent(params.name))
+        ),
     });
   }
```
I'm not opposed to using Promise.all, but it seems less readable and idiomatic for our models to have otherwise-unused consts that are destructured later. Are there good reasons to avoid rsvp.hash?
